### ISSUE TITLE
Feature (optimization) request: add option to disable FF injecting JS and CSS files

### ISSUE THREAD

**Issue by donatasben** [juergenweb/FrontendForms#21]
Hi @juergenweb,
How are you?

I would love to have an option to disable any file(s) that FrontendForms injects into the page html, like

- `mf-conditional-fields.js`
- `frontendforms.js`
- `frontendforms.css`

I am not using any features from these files, use my own custom form CSS (including `.seca`), so these files only take up bandwidth for visitors unnecessarily.

Another suggestion would be for module to selectively load the necessary files depending on enabled/disabled functionality. Like if usage of progress bars are disabled for the form - don't load their JS,CSS files. If no conditional fields are used - do not load their JS file. Etc.

As another optimization I suggest maybe minifying these files.

Or as usually I missed some already existing module option?

**Comment by juergenweb** [comment]
Hello @donatasben 

Nice to hear you, I am fine thanks!

This feature is still implemented but it is a little bit hidden.

![Screenshot 2024-12-12 at 13-58-33 Module FrontendForms • webseite2 at](https://github.com/user-attachments/assets/66277c06-d2ba-4255-b7c9-e5edab6fed8c)

You will find it inside the markup tab section. Take a look at the image. The files can be loaded on each page or on defined pages only, but it is not possible to load only certain features like progressbar,....

Hope this helps

**Comment by donatasben** [comment]
Thanks! Is there a way to do it via API?
Trying `$form->removeCSS(true)` didn't do it for me.

**Comment by juergenweb** [comment]
No, not at the moment. Is there a special need for such a scenario?

**Comment by donatasben** [comment]
Yes, I would like to be able to set this per form.
If page uses several forms and none of them needs these files - none would be loaded.
If at least one form has these enabled - module should load these files (once).

To be honest I don't touch module settings at all and set all settings via API when creating forms. That way I am sure I won't enable/disable via UI something I need for a specific form that I forgot not to touch.
I kind of expected to have every module setting be available to set via API 😁 Would be great!

**Comment by juergenweb** [comment]
Ok, I see, but at the moment I have no really good idea how to realize that. Every form is an own object without connection to each other.
Maybe a solution could be to create a new property of the page called for example frontendforms_js and every form on the page adds a value to this page property (0: no Javascript should be added or 1: Javascript should be added). This results in an array (fe [0,0,1,0] if there are 4 forms on the page). If there at least one 1, then the Javascript file should be added, otherwise not.
But this was only a first thought. If you have a more practicable way, please let me know.

Best regards Jürgen

**Comment by donatasben** [comment]
Yes, that's exactly what I myself already use in my websites to inject JS or CSS selectively!

I set a `page()->set('some_switch', true)` and then when outputting I just check `if (page()->get('some_switch'))`.
The default outcome informs if a boolean is enough, or as you suggest - an array is better to handle possible outcome scenarios.

As you already used it - it's a good idea to use module and file type name space for identifier switch to prevent conflicts.

Finally, when page render hook is firing and FrontendForms is about inject forms - it could check these known switches (or fallback to default/module general setting).

**Comment by juergenweb** [comment]
Thanks for your feedback! I will add this feature in the next update (if it works ;-)

**Comment by juergenweb** [comment]
Hello @donatasben 

I have added your feature requests to the new version 2.2.23. You will find all information inside the [changelog.md](https://github.com/juergenweb/FrontendForms/blob/main/CHANGELOG.md#2223-2024-12-13).

Please test it if everything works as expected.

Have a nice weekend!!

**Comment by juergenweb** [comment]
Version 2.2.24 contains 2 additional performance updates:

- using minified JS and CSS files
- Automatically check whether forms are present on a page or not. If so, then JS and CSS files are injected, otherwise not. It is no longer necessary to manually select the pages in the module configuration on which the files should be loaded. Please note: Embedding these files will only take place if it is not prevented by the module configuration or by the new API methods (useCSS(), useJS()).

So I think that all your suggestions have been implemented, and that's why I'm going to close this request now.

If you discover new issues, please open a new issue report.

Best regards Jürgen

**Comment by donatasben** [comment]
Hi @juergenweb!
Lovely additions! Thank you very much.

But I guess I'm injecting forms into a page in a bit an unusual way (using custom {{placeholder}} implementation - quite similar to the one you implemented) so in my case when the `Page::render` hook fires, the page does not have any forms in it yet and the `addAssets()` function never fires. I was thinking that maybe it would be a good place to inject files straight in the `$form->render();` method? That would make sure that it always is fired.
For example it should load files only when rendering the very first form that uses assets, and any next form could check a global page data param if required files were already included? Or something like that. What do you think? 😃 

I find it as an another page loading optimization - for example if the form is at the very bottom of the page HTML (like a contact form near the footer), the FF asset loading wouldn't block start of browser rendering of the page, and start loading files only when browser encounters the first form in HTML. 😄

**Comment by juergenweb** [comment]
Hello @donatasben 

Am I right, that you are replacing your {{placeholders}} with forms using the Page::render() hook too? So could there be a priority issue in this case, which means that the addAssets() hook runs befor your "placeholder-replace-method"?

In this case it could help to give the addAssets() hook a higher priority, so it will be fired after your replace method for the placeholders.

If this is the case, could you try to add a higher priority (fe 999) to the addAssets() hooked method (100 is the default priority number) - alternatively you can use a lower number than 100 for your replace method.

`$this->addHookAfter('Page::render', $this, 'addAssets', null, ['priority' => '999']);`

Change this code on line 815 in FrontendForms.module file. It would be interesting if this works.

Best regards Jürgen

**Comment by juergenweb** [comment]
Sorry, it must be 

`$this->addHookAfter('Page::render', $this, 'addAssets', ['priority' => '999']);`

Remove the "null" inside the parenthesis.

**Comment by donatasben** [comment]
Hi @juergenweb.
You are right, I use `Page::render` too, and increasing the hook priority in FrontendForms.module solves the problem!

**Comment by juergenweb** [comment]
Ok, I added the value of 999 as a priority to the addAssets hook. This is also a very rare case scenario, so I'm not rewriting the method to put in the render method, and it would make things a lot more complicated. So, increasing the priority is the simplest solution.

At the moment I'm waiting for feedback from another user who discovered a bug in a javascript file and which I just fixed.

If the user gives me the feedback that it works, I will bump up the module to the next version 2.2.25. This version includes the higher priority for the hook too.
